docs/architecture/ADR-003-borrow-assets.md

```diff
@@ -0,0 +1,125 @@
+
+## Detailed Design
+
+For the purposes of borrowing and repaying assets, as well as marking uTokens as collateral, the `umee/x/leverage` module does not mint or burn tokens. It uses its keeper to remember open positions and user collateral settings, and the `x/bank` module to perform all necessary balance checks and token transfers.
+
+Additionally, the `x/bank` module must be extended to prevent users from transferring uTokens when doing so would lower their borrow limit below thir current total borrowed amount.
+
+Because borrow limits weight the value of different token types together, the calculation will require the use of price oracles (and placeholder functions before those are available).
+
+### Basic Message Types
+
+To implement the borrow/repay/collateral functionality of the Asset Facility, the four common message types will be:
+```go
+// MsgSetCollateral - a borrower enables or disables a specific utoken type in their wallet to be used as collateral
+type MsgSetCollateral struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Denom    string         `json:"denom" yaml:"denom"`
+  Enable   bool           `json:"enable" yaml:"enable"`
+}
+// MsgBorrowAsset - a user wishes to borrow assets of one or more allowed types
+type MsgBorrowAsset struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
+}
+// MsgRepayAsset - a user wishes to repay assets of one or more borrowed types
+type MsgRepayAsset struct {
+  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
+  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
+}
+```
+Messages must use only whitelisted denominations. Collateral is always a uToken denomination, and assets are never uTokens.
+
+Asset borrowing respects the user's borrowing limit, which compares collateral uTokens associated with one base asset, to borrowed assets of a second base asset type. Price oracles must be used to compare the values of base assets (e.g. Atoms:Ether.)
+
+_Note: The `Coins` type seen in the `Amount` fields can contain multiple token types. The messages above should fail if even one of the coin types fails or is invalid, rather than partially succeeding._
+
+It is necessary that messages be signed by the borrower's account. Thus the method `GetSigners` should return the `Borrower` address for all message types above.
+
+### Keeper additions
+
+Using the `sdk.Coins` built-in type, which combines multiple {Denom,Amount} pairs as a single object, the `umee/x/leverage` keeper may roughly remember open borrow and collateral positions as follows:
+
+```go
+// pseudocode
+// for each borrower address, combine all open borrow positions into one sdk.Coins object:
+keeper[borrowPrefix + borrowerAddress] = BorrowedCoins
+// additionally user collateral settings are stored as a slice of enabled denoms
+keeper[collateralPrefix + borrowerAddress] = []string{denom1,denom2...}
```

Comment:
This is one way we could store user collateral settings (we need to know the list of uToken denoms they've enabled).

We could also avoid using []string by using an sdk.Coins with an sdk.Coin (amount = 1 or some placeholder value) for each enabled denom.

Reply:
I would store this as follows, and in general recommend this approach to annotating how storage keys are defined in ADRs:

```
collateralPrefix | lengthPrefixed(lenderAddress) | tokenDenom
```

where `|` denotes concatenation

Reply:
Let's update both key formats in this ADR 👍